### PR TITLE
GH-3395: Use Original Key in Reply Record

### PR DESCRIPTION
**Pull request by sobychacko**
Fixes: #3395

When using `ReplyingKafkaTemplate,` include the original key from the request record if such a key exists.

### CODE REVIEW

File: spring-kafka-docs/src/main/antora/modules/ROOT/pages/kafka/sending-messages.adoc [hunk 1]
```diff
@@ -650,6 +650,13 @@ public Message<?> messageReturn(String in) {
 }
 ----
 
+[[record-key-in-reply]]
+== Preserving Request Record Key in Reply
```
Comment (artembilan) [inline spring-kafka-docs/src/main/antora/modules/ROOT/pages/kafka/sending-messages.adoc:654]: Why do we need a dedicated section for this single short paragraph?
Why it cannot be included in the end if the previous one?
At least it looks like both of them talks about similar context.
Where that `Message` indeed covers what you try to say about multi-value reply.

File: spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java [hunk 2]
```diff
@@ -638,12 +638,12 @@ private void sendReplyForMessageSource(Object result, String topic, Message<?> s
 			builder.setHeader(this.correlationHeaderName, correlationId);
 		}
 		setPartition(builder, source);
+		setKey(builder, source);
 		this.replyTemplate.send(builder.build());
 	}
 
 	protected void asyncSuccess(@Nullable Object result, String replyTopic, Message<?> source,
 			boolean returnTypeMessage) {
-
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java:646]: This is really a good style to make code more readable.
There is might no reason to do that for single-like method header, but if it is multi-line, then better to separate body for easy review.
Please, revert this change and fix your IDE to preserver such a blank line on refactoring.

File: spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java [hunk 3]
```diff
@@ -719,11 +717,24 @@ private void setPartition(MessageBuilder<?> builder, Message<?> source) {
...
+	}
+
 	@Nullable
 	private byte[] getReplyPartition(Message<?> source) {
 		return source.getHeaders().get(KafkaHeaders.REPLY_PARTITION, byte[].class);
 	}
 
+	@Nullable
+	private Object getReplyKeyFromRequest(Message<?> source) {
+		return source.getHeaders().get(KafkaHeaders.RECEIVED_KEY);
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java:735]: I don't see a reason in this one-line method.
More over it feel like `static`.
Way, either way: I'd prefer to not have it making me thinking "where else this method can be used" 😄

**Review by artembilan — CHANGES_REQUESTED** [review]

File: spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java [hunk 4]
```diff
@@ -662,7 +663,6 @@ protected void acknowledge(@Nullable Acknowledgment acknowledgment) {
 
 	protected void asyncFailure(Object request, @Nullable Acknowledgment acknowledgment, Consumer<?, ?> consumer,
 			Throwable t, Message<?> source) {
-
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java:665]: Why haven't you reverted this?

File: spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java [hunk 5]
```diff
@@ -676,7 +676,6 @@ protected void asyncFailure(Object request, @Nullable Acknowledgment acknowledgm
 
 	protected void handleException(Object records, @Nullable Acknowledgment acknowledgment, Consumer<?, ?> consumer,
 			Message<?> message, ListenerExecutionFailedException e) {
-
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java:679]: DITTO, please.

File: spring-kafka-docs/src/main/antora/modules/ROOT/pages/kafka/sending-messages.adoc [hunk 6]
```diff
@@ -650,6 +650,12 @@ public Message<?> messageReturn(String in) {
 }
 ----
 
+=== Original Record Key in Reply
```
Comment (artembilan) [inline spring-kafka-docs/src/main/antora/modules/ROOT/pages/kafka/sending-messages.adoc:653]: Why do you still think that this short paragraph deserves its own title?

**Review by artembilan — CHANGES_REQUESTED** [review]

File: spring-kafka-docs/src/main/antora/modules/ROOT/pages/kafka/sending-messages.adoc (hunk 6 above)
Comment (sobychacko) [reply spring-kafka-docs/src/main/antora/modules/ROOT/pages/kafka/sending-messages.adoc:653]: I don't see a problem with a separate section for this, but I am neutral either way.
